src/client/datascience/ipywidgets/commonMessageCoordinator.ts

```diff
@@ -236,11 +241,20 @@ export class CommonMessageCoordinator {
                 this.serviceContainer.get<IPersistentStateFactory>(IPersistentStateFactory),
                 this.serviceContainer.get<IExtensionContext>(IExtensionContext)
             );
-            this.disposables.push(this.ipyWidgetScriptSource.postMessage(this.postEmitter.fire.bind(this.postEmitter)));
-            this.disposables.push(
-                this.ipyWidgetScriptSource.postInternalMessage(this.postEmitter.fire.bind(this.postEmitter))
-            );
+            this.disposables.push(this.ipyWidgetScriptSource.postMessage(this.cacheOrSend, this));
+            this.disposables.push(this.ipyWidgetScriptSource.postInternalMessage(this.cacheOrSend, this));
         }
         return this.ipyWidgetScriptSource;
     }
+    private cacheOrSend(data: any) {
+        // If no one is listening to the messages, then cache these.
+        // It means its too early to dispatch the messages, we need to wait for the event handlers to get bound.
+        if (!this.listeningToPostMessageEvent) {
```

Comment:
I personally think we should refactor all of this, this is just too clunky, however we have no choice because there's a lot of async stuff (it all goes down to getting the INotebook which happens to be async)